### DISCUSSION TITLE
buildTarget vs dependsOn

### DISCUSSION THREAD

**Discussion by vicb** [nrwl/nx#14243] [General]
I am working on implementing the preview server for vite (#14125).

In this context it's not clear to me when to use `buildTarget` vs `dependsOn`.

Looking at the node plugin:

```
    "serve": {
      "executor": "@nrwl/js:node",
      "options": {
        "buildTarget": "fxc-server:build"
      },
      "configurations": {
        "production": {
          "buildTarget": "fxc-server:build:production"
        }
      }
    },
```

The `buildTarget` effectively [adds a dependency](https://github.com/nrwl/nx/blob/ef55bf776e78d302acef86b465d860ce8d1261b8/packages/js/src/executors/node/node.impl.ts#L196-L213) on the build.

There are a few questions raised here:

1) When to use `buildTarget` vs `dependsOn`? It seems like one of the difference is that you can specify a configuration in `buildTarget` but you can not for `dependsOn`

2) Would it make sense to add the possibility to pass a configuration to `dependsOn`? I think it still would not solve all the problems as `dependsOn` is at the target level (vs `buildTarget` at the configuration level).

3) Would it make sense to be able to programmatically add a dependency? That is the executor could add a dependency on the build target. That way we do not need to explicitly run the dependency and also the dependency is known to Nx (and could appear on the task graph).  

4) Semi-(un)related question. Is it expected that if target tA depends on target tB and target tA has no configurations then you can `nx project:tA --configuration=<any of target B configuration>`. See https://github.com/nrwl/nx-console/issues/1466.

Thank you for Nx, eager to know more about the design.

/cc @mandarini @vsavkin

**Comment by nartc** [comment]
Hi, thank you for your effort on trying to contribute to Nx. I can chime in here since you used `nrwl/js:node` as an example.

1. `buildTarget` vs `dependsOn`. These two serve different purposes. `dependsOn` makes sure an upstream target (or targets) have to finish **successfully** before Nx can execute a current target, usually in tandem with Caching mechanism and Cache Invalidation. Eg: `test` has `dependsOn: ['build']`. This is saying, we only want to `test` if `build` succeeds. It doesn't **necessarily** mean that `test` will use `build`'s artifacts (eg: the `dist/*` directory). On the other hand, `buildTarget` is a custom property for a particular Executor (eg: in this case, `nrwl/js:node`) and is used to invoke **AND** listen for the `buildTarget`'s result, then use this result to do something else. For `nrwl/js:node`, we don't just say "run `build` before run `serve`", we do:
- If `build` succeeds, run the entry point in `build`'s `outputPath`. 
- If `build` restarts, live/hot reload
- If `build` errors, logs the errors

To give a quick answer to the question of "When to use `dependsOn` vs `buildTarget`", it depends on what target, or more accurately the underlying executor, is trying to do. Does it need to use the result's of the parent target/executor? Does it need to live/hot reload? A quick deciding factor is _serve_ targets should almost always have `buildTarget`. Should `preview` do that? Should `preview` reload the `build` artifacts on changes? 

2. Can you give a more concrete example of what "passing a configuration to `dependsOn`" is trying to solve? I personally don't think it makes much sense but would love to understand more your stand point.

3. Sorry but I don't think I fully understand this question/feature request. Maybe someone can chime in

4. I read the issue over at `nx-console` and I think `nx` is working as expected. I have the following dummy setup

```json
{
    "targets": {
        "build": {
            "options": {
                "outputPath": "dist/project"
            },
            "configurations": {
                "production": {
                    "outputPath": "dist/project/prod"
                }
            }
        },
        "dummy": {
            "options": {
                "command": "echo hello"
            },
            "dependsOn": ["build"]
        }
    }
}
```

- Running `npx nx dummy project` runs
    - `npx nx run project:build` (without any configuration)
    - `npx nx run project:dummy` (finally logs "hello" here)
- Running `npx nx dummy project --configuration=production` runs:
    - `npx nx run project:build:production` because `build` has `production` configuration
    - `npx nx run project:dummy` (without any configuration) logs "hello"

As you can see here, `project:dummy` doesn't get invoked with some "implicit" `production` configuration. Nx simply runs `project:dummy` with the **default** `options` set. So there's really no unexpected behavior here. To your point/question, if you want to use the Nx Console to invoke `fetcher` with `build:production` then you can add `defaultConfiguration: 'production'` to `build`. But this is not 100% ideal because other targets that depends on `build` with the default `options` are need to adjust. 

Again, thank you for the discussion and I hope I was able to clear up some points. Feel free to ask further questions.

**Comment by vicb** [comment]
Thanks for your detailed answer @nartc.

That along with looking more deeply at the code helped me.

1) `dependsOn` vs `buildTarget`

I think the best we can do for now is:
- Use `dependsOn` if we **only** need to run after the dependency has executed successfully,
- Use `buildTarget` if we need to do more (i.e. consuming the dependency output, iterate the dependency generator, ...).

That's how I'll use it.

I still think not using `dependsOn` might cause troubles down the line.

As an example when `buildTarget` is used to run an executor (i.e. `nrwl/js:node`) there is no dependency known to Nx on the referenced build target (i.e. `serve` does not depend on `build`) and it messes with the [`Task Graph`](https://nx.dev/core-features/distribute-task-execution#running-things-in-parallel). I think you can end up in a situation where the same target is executed multiple times concurrently - not sure how that would play with the cache as this situation is not supposed to happen in the first place. Also note that because you can specify the configuration for the `buildTarget` you might end up executing different configuration simultaneously.

Hopefully `buildTarget` is not widespread.

2) When it would make sense to pass a config to `dependsOn`.

See my example from 4)

```
    "container": {
      "executor": "@nx-tools/nx-container:build",
      "dependsOn": ["build"],
      "options": {
        // ...
      },
    }
```

I really **never** want to build a Docker image based on my development configuration. Being able to specify `build:production` would avoid shooting myself in the foot.

An other example would be a `deploy` target with `production` and `qa` configurations. If you run this target in the `qa` configuration then you would depend on `build[:default]` we might not be what you want.

3) programmatically add a dependency?

My idea was to keep the Task Graph valid by adding a dependency when you use a `buildTarget`.
Thinking more about it I think the solution (long term) is enhancing `dependsOn` so that you can use that to run the target and never have to launch another executor from your executor code.

4) config propagation

> then you can add defaultConfiguration: 'production' to build

Changing the `defaultConfiguration` impact all the dependent tasks.
It could also have a wider impact if you consider what `nx run-many` or `nx affected` will execute.
